### PR TITLE
Introduce number of samples to TensorListShape

### PR DESCRIPTION
**Pull request by klecki**
Allow for 0-element TLS.
Fixes size for 0-dim TLS.

Add tests for size().
Add non-template overload so set_tensor_shape
can be used with brace initializer.

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/kernels/tensor_shape.h [hunk 1]
```diff
@@ -495,6 +495,10 @@ struct TensorListShapeBase {
     return out;
   }
 
+  void set_tensor_shape(int64_t sample, const TensorShape<sample_ndim> &sample_shape) {
```
Comment (mzient) [inline dali/kernels/tensor_shape.h]: Why is this necessary?

File: dali/kernels/tensor_shape.h [hunk 2]
```diff
@@ -612,7 +640,7 @@ struct TensorListShape<DynamicDimensions>
   }
 
   int sample_dim() const { return dim; }
-  int size() const { return shapes.size() / sample_dim(); }
+
   void set_sample_dim(int dim) { this->dim = dim; }
```
Comment (mzient) [inline dali/kernels/tensor_shape.h]: A side note:
How about making it private and befriending `TensorListShape` with its base class? This function leaves the object in an inconsistent state. I know that our data members our public, but that's part of our contract - sort of "you modify object directly, better know what you're doing" thing. Public functions, on the other hand, are expected to keep object usable.

File: dali/kernels/tensor_shape.h (hunk 1 above)
Comment (klecki) [reply dali/kernels/tensor_shape.h]: This is to allow `tls.set_tensor_shape(i, {1, 2, 3});` without compiler complaining for ambiguity. And I am not keen on using initializer_list as it has weird rules of promotions.

File: dali/kernels/tensor_shape.h (hunk 2 above)
Comment (klecki) [reply dali/kernels/tensor_shape.h]: Done. Good idea.

**Comment by klecki** [comment]
!build

File: dali/kernels/tensor_shape.h (hunk 1 above)
Comment (mzient) [reply dali/kernels/tensor_shape.h]: Fair enough.

File: dali/kernels/tensor_shape.h [hunk 3]
```diff
@@ -768,6 +809,9 @@ bool operator==(const TensorListShape<left_ndim> &left, const TensorListShape<ri
   if (left.sample_dim() != right.sample_dim()) {
```
Comment (mzient) [inline dali/kernels/tensor_shape.h]: I wonder if we should allow comparison of two fixed-sample-dim TensorListShapes. Maybe add:
```
detail::check_compatible_ndim<left_ndim, right_ndim>();
```

Comment (klecki) [reply dali/kernels/tensor_shape.h]: Done

File: dali/kernels/tensor_shape.h (hunk 1 above)
Comment (klecki) [reply dali/kernels/tensor_shape.h]: Done

**Comment by klecki** [comment]
!build

File: dali/kernels/tensor_shape.h [hunk 4]
```diff
@@ -510,6 +515,7 @@ struct TensorListShapeBase {
 
   decltype(shapes.data()) data() { return shapes.data(); }
   constexpr bool empty() const { return size() == 0; }
+  int size() const { return num_samples_; }
```
Comment (JanuszL) [inline dali/kernels/tensor_shape.h]: Do we need `size` if `num_samples` is already here?

Comment (klecki) [reply dali/kernels/tensor_shape.h]: Which one should we use?
TensorShape and TensorListShape have size() for number of elements and number of samples correspondingly.
Tensor and TensorList use size() from dali::Buffer to return the number of elements (allocation).
And TensorList has ntensor() for obtaining the number of samples. Should we unify that somehow?

**Comment by klecki** [comment]
!build

File: dali/kernels/tensor_shape.h (hunk 4 above)
Comment (klecki) [reply dali/kernels/tensor_shape.h]: If we decide on something, I will do a follow-up.

File: dali/kernels/tensor_shape.h [hunk 5]
```diff
@@ -645,23 +696,51 @@ struct TensorListShape : TensorListShapeBase<TensorListShape<sample_ndim>, sampl
...
-  TensorListShape(TensorListShape &&) = default;
+  TensorListShape(TensorListShape &&other) : Base(std::move(other.shapes), other.num_samples_) {
+    other.num_samples_ = 0;
+  }
+
+  explicit TensorListShape(int num_samples) : Base() {
+    Base::resize(num_samples);
+  }
+
+  TensorListShape(int num_samples, int ndim) : Base()  {
```
Comment (mzient) [inline dali/kernels/tensor_shape.h]:
```suggestion
  explicit TensorListShape(int num_samples, int ndim) : Base()  {
```

File: dali/kernels/tensor_shape.h [hunk 6]
```diff
@@ -645,23 +696,51 @@ struct TensorListShape : TensorListShapeBase<TensorListShape<sample_ndim>, sampl
...
+    assert(ndim == sample_ndim);
+    Base::resize(num_samples, ndim);
+  }
 
   TensorListShape(const std::vector<TensorShape<sample_ndim>> &sample_shapes)  // NOLINT
-      : Base(flatten_shapes(sample_shapes)) {}
+      : Base(flatten_shapes(sample_shapes), sample_shapes.size()) {}
 
   TensorListShape(const std::vector<int64_t> &shapes, int ndim)
```
Comment (mzient) [inline dali/kernels/tensor_shape.h]:
```suggestion
  TensorListShape(const std::vector<int64_t> &shapes, int sample_dim)
```

File: dali/kernels/tensor_shape.h [hunk 7]
```diff
@@ -645,23 +696,51 @@ struct TensorListShape : TensorListShapeBase<TensorListShape<sample_ndim>, sampl
...
-      : Base(shapes) {
+      : Base(shapes, shapes.size() / ndim) {
     assert(ndim == sample_ndim);
   }
 
   TensorListShape(std::vector<int64_t> &&shapes, int ndim)
```
Comment (mzient) [inline dali/kernels/tensor_shape.h]:
```suggestion
  TensorListShape(std::vector<int64_t> &&shapes, int sample_dim)
```

File: dali/kernels/tensor_shape.h [hunk 8]
```diff
@@ -574,35 +588,69 @@ struct TensorListShape<DynamicDimensions>
...
-      : Base(std::move(other.shapes)), dim(other.sample_dim()) {}
+      : Base(std::move(other.shapes), other.size()), dim(other.sample_dim()) {
+    other.num_samples_ = 0;
+  }
+
+  explicit TensorListShape(int num_samples) : Base() {
+    resize(num_samples);
+  }
+
+  TensorListShape(int num_samples, int ndim) : Base()  {
```
Comment (mzient) [inline dali/kernels/tensor_shape.h]:
```suggestion
  explicit TensorListShape(int num_samples, int ndim) : Base()  {
```

File: dali/kernels/tensor_shape.h [hunk 9]
```diff
@@ -574,35 +588,69 @@ struct TensorListShape<DynamicDimensions>
...
-      : Base(flatten_shapes(sample_shapes)), dim(get_dim_from_uniform(sample_shapes)) {}
+      : Base(flatten_shapes(sample_shapes), sample_shapes.size()),
+        dim(get_dim_from_uniform(sample_shapes)) {}
 
   TensorListShape(const std::vector<TensorShape<DynamicDimensions>> &sample_shapes)  // NOLINT
-      : Base(flatten_shapes(sample_shapes)), dim(get_dim_from_uniform(sample_shapes)) {}
+      : Base(flatten_shapes(sample_shapes), sample_shapes.size()),
+        dim(get_dim_from_uniform(sample_shapes)) {}
+
+  TensorListShape(const std::vector<int64_t> &shapes, int ndim)
```
Comment (mzient) [inline dali/kernels/tensor_shape.h]:
```suggestion
  TensorListShape(const std::vector<int64_t> &shapes, int sample_dim)
```

File: dali/kernels/tensor_shape.h [hunk 10]
```diff
@@ -574,35 +588,69 @@ struct TensorListShape<DynamicDimensions>
...
+      : Base(shapes, shapes.size() / ndim), dim(ndim) {}
+
+  TensorListShape(std::vector<int64_t> &&shapes, int ndim)
```
Comment (mzient) [inline dali/kernels/tensor_shape.h]:
```suggestion
  TensorListShape(std::vector<int64_t> &&shapes, int sample_dim)
```

File: dali/kernels/tensor_shape.h (hunk 5 above)
Comment (klecki) [reply dali/kernels/tensor_shape.h]: Done

File: dali/kernels/tensor_shape.h (hunk 6 above)
Comment (klecki) [reply dali/kernels/tensor_shape.h]: Done, all around the file

File: dali/kernels/tensor_shape.h (hunk 8 above)
Comment (klecki) [reply dali/kernels/tensor_shape.h]: Done

**Comment by klecki** [comment]
!build
